docs/architecture.md

```diff
@@ -1,32 +1,139 @@
+Cortex has a fundamentally service-based architecture, in which the overall system is split up into a variety of components that perform specific tasks and run separately (and potentially in parallel).
 
-See the document "[Cortex Query Woes](https://docs.google.com/document/d/1lsvSkv0tiAMPQv-V8vI2LZ8f4i9JuTRsuPI_i-XcAqY)" for more details design discussion. In the future, query splitting, query alignment and query results caching will be added to the frontend.
+### Distributor
 
-The query frontend is completely optional - you can continue to use the queriers directly.  If you want to use the query frontend, direct incoming authenticated traffic at them and set the `-querier.frontend-address` flag on the queriers.
+The **distributor** service is responsible for handling samples written by Prometheus. It's essentially the "first stop" in the write path for Prometheus samples, *unless* you run an optional [query frontend](#query-frontend), in which case the distributor takes writes from the query frontend rather than from Prometheus instances directly. Once the distributor receives samples from Prometheus, it splits them into chunks, replicates them (based on a configurable replication factor), and then sends them to the [ingester](#ingester) service.
 
-### Queriers
+Distributors communicate with ingesters via [gRPC](https://grpc.io). They are stateless and can be scaled up and down as needed.
 
-The queriers handled the actual PromQL evaluation.  They embed the chunk store client code for fetching data from long-term storage, and communicate with the ingesters for more recent data.
+#### Hashing
+
+Distributors use consistent hashing, in conjunction with the (configurable) replication factor, to determine *which* instance of the ingester service receives each sample. The hash itself is based on the metric name and tenant ID for the series. Cortex ensures that the samples for a given series all end up on the same subset of ingesters.
+
+> This hashing scheme was originally chosen to reduce the number of required ingesters on the query path. The trade-off, however, is that the write load on the ingesters is less even.
+
+#### The hash ring
+
+A consistent hash ring is stored in [Consul](https://www.consul.io/) as a single key-value pair, with the ring data structure also encoded as a [Protobuf](https://developers.google.com/protocol-buffers/) message. The consistent hash ring consists of a list of tokens and ingesters. Hash values are looked up in the ring; the replication set is built for the closest unique ingesters by token. One of the benefits of this system is that adding and remove ingesters results in only 1/_n_ of the series being moved (where _n_ is the number of ingesters).
+
+All distributors share access to the same hash ring, which means that write requests can be sent to any distributor.
```

Comment:
This sentence belong with the #### Load balancing across distributors section IMO.